onnxruntime/core/optimizer/attention_fusion.cc

```diff
@@ -558,6 +453,139 @@ bool AttentionFusion::FuseSubGraph(Node& layer_norm, const Node& add_after_layer
+          |          Softmax       /                                    |
+          |             \         /                                     |
+          |              \       /                                      |
+          |            qkv_MatMul                                       |
+          |                   |                                         |
+          |                Transpose (perm=0,2,1,3)                     |
+          |                   |                                         |
+          |                Reshape-----------------------------------Concat [Shape=_,-1,W]
+          |                   |
+          |                 MatMul----[Weights](WxW)
+          |                   |
+          |                  Add----[Bias](W)
+          +-------------------|---+
+                              |   |
+                               Add
+
+Two changes compared with first version attention fusion for distilbert:
+1:TransposeScaleMatmul[alpha = 0.125, transA = 0, transB = 0] is initially Div[B = 8] -> MatMul. It gets fused before ORT 1.5.0 release
+2:There were two Shape nodes after LayerNormalization, gets fused into one before ORT 1.5.0 release
+
+However, the first version of attention fusion for distilbert is still supported for now.
+*/
+static bool FuseSubGraphQKDistilBert(Node& layer_norm,
+                                     Graph& graph,
+                                     AttentionFusionHelper::AttentionMaskNodesDistilBert& mask_nodes,
+                                     NodeArg* mask_input,
+                                     std::vector<std::reference_wrapper<const Node>>& parent_path_nodes,
+                                     int64_t hidden_size,
+                                     int64_t num_heads,
+                                     int64_t head_size,
+                                     std::map<std::string, NodeArg*>& mask_index_map,
+                                     const logging::Logger& logger) {
+  // path to q
+  std::vector<graph_utils::EdgeEndToMatch> q_varience_path{
+      {0, 2, "MatMul", {1, 9, 13}, kOnnxDomain},
+      {0, 0, "Div", {7, 13}, kOnnxDomain}};
+  std::vector<const Node::EdgeEnd*> edges;
+  if (!graph_utils::FindPath(*(mask_nodes.where), true, q_varience_path, edges, logger)) {
+    DEBUG_LOG("Failed to find path for q");
+    return false;
+  }
+
+  std::vector<NodeIndex> nodes_to_remove;
+  if (!FuseSubGraphQKImpl(layer_norm, graph, parent_path_nodes, mask_input, mask_index_map, edges, nodes_to_remove, hidden_size,
+                          num_heads, head_size, logger)) {
+    return false;
+  }
+
+  const Node& reshape_1 = parent_path_nodes[0];
+  const Node& reshape_2 = *(mask_nodes.reshape);
+
+  const Node* p_concat_1 = graph_utils::GetInputNode(reshape_1, 1);
+  const Node* p_concat_2 = graph_utils::GetInputNode(reshape_2, 1);
+  if (p_concat_1 != nullptr && p_concat_2 != nullptr) {
```

Comment:
Will them be NULL? what if p_concat_1 is nullptr and p_concat_2 is not nullptr, then some nodes might not be removed?

Reply:
they should not be null in this case. Will add else {return false}